runtime/tests/checker/interface_test.go

```diff
@@ -1991,3 +1992,538 @@ func TestCheckInvalidInterfaceUseAsTypeSuggestion(t *testing.T) {
+              return Test.X().test()
+          }
+        `)
+		require.NoError(t, err)
+	})
+}
+
+func TestCheckInterfaceDefaultImplementationOverriden(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("interface", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          struct interface IA {
+              fun test(): Int {
+                  return 41
+              }
+          }
+
+          struct Test: IA {
+              fun test(): Int {
+                  return 42
+              }
+          }
+
+          fun test(): Int {
+              return Test().test()
+          }
+        `)
+		require.NoError(t, err)
+	})
+
+	t.Run("type requirement", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          contract interface IA {
+
+              struct X {
+                  fun test(): Int {
+                      return 41
+                  }
+              }
+          }
+
+          contract Test: IA {
+
+              struct X {
+                  fun test(): Int {
+                      return 42
+                  }
+              }
+          }
+
+          fun test(): Int {
+              return Test.X().test()
+          }
+        `)
+		require.NoError(t, err)
+	})
+}
+
+func TestCheckInvalidInterfaceDefaultImplementationConcreteTypeUsage(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("interface", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          struct interface IA {
+              fun test(): Int {
+                  return self.x
+              }
+          }
+
+          struct Test: IA {
+              let x: Int
+
+              init() {
+                  self.x = 0
+              }
+          }
+
+          fun test(): Int {
+              return Test().test()
+          }
+        `)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		require.IsType(t, &sema.NotDeclaredMemberError{}, errs[0])
+	})
+
+	t.Run("type requirement", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          contract interface IA {
+
+              struct X {
+                  fun test(): Int {
+                      return self.x
+                  }
+              }
+          }
+
+          contract Test: IA {
+
+              struct X {
+                  let x: Int
+
+                  init() {
+                      self.x = 0
+                  }
+              }
+          }
+
+          fun test(): Int {
+              return Test.X().test()
+          }
+        `)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		require.IsType(t, &sema.NotDeclaredMemberError{}, errs[0])
+	})
+
+}
+
+func TestCheckInterfaceDefaultImplementationConcreteTypeUsage(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("interface", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          struct interface IA {
+              let x: Int
+
+              fun test(): Int {
+                  return self.x
+              }
+          }
```

Comment:
I'm more interested in the runtime behavior of this example. How would the inherrited function pick the correct variable scope at runtime?

Probably good to add as a test case?

Reply:
Hey @SupunS  sorry for the late answer, I could just get back into this, I will add a test case. 

But I am not too versed in variable scopes in Cadence tbh, so any help or clarification is much appreciated.

Reply:
Hey @bluesign , Sorry for being so late to get back on this. I went through the code once more, and I think now I understand how the scoping work with the default functions. The [defaultFunctions wrapper method](https://github.com/onflow/cadence/blob/6c92d06d9a6261797a0cf2edbd142bdb7f3cce78/runtime/interpreter/interpreter.go#L2000) always takes the current scope/variableActivation.

I added a test-case too: https://github.com/onflow/cadence/commit/eb899b0fca520c7aa93823ed54631a5ec63cd823 (this is in a forked branch) to verify the behaviour. I can push it to this PR, if that's OK?

Reply:
@SupunS sure thanks a lot